src/dmd/lexer.d

```diff
@@ -202,6 +203,104 @@ unittest
     }
 }
 
+
+/**************
+ * Pool allocator for Token objects, used by the Lexer.
+ * Tokens are never freed so Token pointers will never change
+ */
+private struct TokenPool
+{
+    private Array!Token tokens;
+
+    public ref Token getByIndex(size_t index) nothrow pure
+    {
+        if (index >= tokens.length)
+        {
+            tokens.setDim(index + 1);
+        }
+        return tokens[index];
+    }
+}
+
+
+/***************
+ * Represents a range of tokens lazily scanned from a lexer
+ * should be created with the Lexer makeRangeFromHere method
+ */
+struct TokenRange
+{
+private:
```

Comment:
I recommend using `private` in front of each instance variable instead and remove `public:`.